controller/tap/server.go

```diff
@@ -469,3 +485,68 @@ func NewServer(
 
 	return s, lis, nil
 }
+
+func indexPodByIP(obj interface{}) ([]string, error) {
+	if pod, ok := obj.(*apiv1.Pod); ok {
+		return []string{pod.Status.PodIP}, nil
+	}
+	return []string{""}, fmt.Errorf("object is not a pod")
+}
+
+// hydrateIPMeta returns a map of expanded metadata labels for a given IP.
+// If no pod cound be found for that IP, it returns an empty map.
+func (s *server) hydrateIPMeta(ip *public.IPAddress) (map[string]string, error) {
+	pod, err := s.podForIP(ip)
+	if err != nil {
+		return nil, err
+	}
+	if pod == nil {
+		log.Debugln("no pod for IP %s", addr.PublicIPToString(ip))
+		return make(map[string]string), nil
+	}
+
+	ownerKind, ownerName := s.k8sAPI.GetOwnerKindAndName(pod)
+	return pkgK8s.GetPodLabels(ownerKind, ownerName, pod), nil
+}
+
+// podForIP returns the pod corresponding to a given IP address, if one exists.
+//
+// If multiple pods exist with the same IP address, this may be because some
+// are terminating and the IP has been assigned to a new pod. In this case, we
+// select the running pod, if one currently exists. If there is a single pod
+// which is not running, we return that pod. Otherwise we return `nil`, as we
+// cannot easily determine which pod sent a given request.
+//
+// If no pods were found for the provided IP address, it returns nil. Errors are
+// returned only in the event of an error indexing the pods list.
+func (s *server) podForIP(ip *public.IPAddress) (*apiv1.Pod, error) {
+	ipStr := addr.PublicIPToString(ip)
+	objs, err := s.k8sAPI.Pod().Informer().GetIndexer().ByIndex(podIPIndex, ipStr)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(objs) == 1 {
+		log.Debugf("found one pod at IP %s", ipStr)
+		return objs[0].(*apiv1.Pod), nil
+	}
+
+	for _, obj := range objs {
+		// It's safe to cast the object to a pod here, as otherwise, it would
+		// not have been indexed by the indexing func in the first place.
+		pod := obj.(*apiv1.Pod)
```

Comment:
I don't think it's worth commenting about the cast here. You're already casting earlier in this method on line 532 without a comment, so it feels inconsistent to call it out here.